File: python/backtracking/magnet_puzzle.py
class MagnetPuzzle:
    """Class to solve the magnet puzzle solution"""

    def __init__(self, size, top, bottom, left, right, slots):
        """
        Constructor which takes input values
        Parameters
        size   - the size of the board in the form of a tuple
        top    - array with entries corresponding to '+' in columns
        bottom - array with entries corresponding to '-' in columns
        left   - array with entries corresponding to '+' in rows
        right  - array with entries corresponding to '-' in rows
        slots  - matrix that stores the data of slots LR for horizontal slot TB for vertical slot
        """

        self.M, self.N = size
        self.top, self.bottom, self.left, self.right = top, bottom, left, right
        self.slots = slots
        self.board = [['X' for j in range(N)] for i in range(M)]

    
    def checkConstraints(self, row, col, ch):
        """
        Function to check if it safe to put 'ch' at board[row][col]
        Parameters
        row, col - cell where the chracter ch is to put
        ch       - the character to be put
        """
        
        # Check for contents of orthogonal adjacent cells to ensure they don't contain same character
        board = self.board
        top, bottom = row - 1 >= 0 and board[row - 1][col] == ch, row + 1 < M and board[row + 1][col] == ch
        left, right = col - 1 >= 0 and board[row][col - 1] == ch, col + 1 < N and board[row][col + 1] == ch
        if top or right or bottom  or left:
            return False

        # If given character is '+', check top & left if given character is '-', check bottom & right
        top, bottom, left, right = self.top, self.bottom, self.left, self.right
        rowCount = sum([board[row][j] == ch for j in range(self.N)])
        colCount = sum([board[i][col] == ch for i in range(self.M)])
        if ch == '+':
            if (top[col] != -1 and colCount >= top[col]) or (left[row] != -1 and rowCount >= left[row]):
                return False
        elif ch == '-':
            if (bottom[col] != -1 and colCount >= bottom[col]) or (right[row] != -1 and rowCount >= right[row]):
                return False
        
        return True

    
    def validate(self):
        """
        Function to validate Configuration of output board
        """
        
        def countInRow(ch, i):
            return sum([self.board[i][j] == ch for j in range(self.N)])
        for i in range(self.M):
            if self.left[i] not in [-1, countInRow('+', i)] or self.right[i] not in [-1, countInRow('-', i)]:
                return False
                
        def countInColumn(ch, j):
            return sum([self.board[i][j] == ch for i in range(self.M)])
        for j in range(self.N):
            if self.top[j] not in [-1, countInColumn('+', j)] or self.bottom[j] not in [-1, countInColumn('-', j)]:
                return False
    
        return True

    def solve(self, row, col):
        """
        Function that solves the board given the initial conditions
        Parameters
        row - starting cell row index
        col - starting cell column index
        """
        
        top, bottom, left, right = self.top, self.bottom, self.left, self.right
        slots, board = self.slots, self.board
        checkConstraints = self.checkConstraints
        # if we have reached last cell
        if row >= self.M - 1 and col >= self.N - 1:
            return self.validate()

        # if last column of current row is already processed, go to next row, first column
        if col >= self.N:
            col, row = 0, row + 1

        # if current cell contains R (end of horizontal slot) or B (end of vertical slot) recur for next cell
        if slots[row][col] == 'R' or slots[row][col] == 'B':
            if self.solve(row, col + 1):
                return True

        # if horizontal slot contains L and R
        if slots[row][col] == 'L' and slots[row][col + 1] == 'R':
            # put ('+', '-') pair and recur
            if checkConstraints(row, col, '+') and checkConstraints(row, col + 1, '-'):
                board[row][col] = '+'
                board[row][col + 1] = '-'
                if self.solve(row, col + 2):
                    return True
                # if it doesn't lead to a solution, backtrack
                board[row][col] = 'X'
                board[row][col + 1] = 'X'

            # put ('-', '+') pair and recur
            if checkConstraints(row, col, '-') and checkConstraints(row, col + 1, '+'):
                board[row][col] = '-'
                board[row][col + 1] = '+'
                if self.solve(row, col + 2):
                    return True
                # if it doesn't lead to a solution, backtrack
                board[row][col] = 'X'
                board[row][col + 1] = 'X'

        # if vertical slot contains T and B
        if slots[row][col] == 'T' and slots[row + 1][col] == 'B':
            # put ('+', '-') pair and recur
            if checkConstraints(row, col, '+') and checkConstraints(row + 1, col, '-'):
                board[row][col] = '+'
                board[row + 1][col] = '-'
                if self.solve(row, col + 1):
                    return True
                # if it doesn't lead to a solution, backtrack
                board[row][col] = 'X'
                board[row + 1][col] = 'X'

            # put ('-', '+') pair and recur
            if checkConstraints(row, col, '-') and checkConstraints(row + 1, col, '+'):
                board[row][col] = '-'
                board[row + 1][col] = '+'
                if self.solve(row, col + 1):
                    return True
                # if it doesn't lead to a solution, backtrack
                board[row][col] = 'X'
                board[row + 1][col] = 'X'

        # ignore current cell and recur
        if self.solve(row, col + 1):
            return True

        # if no solution is possible, return false
        return False


    def display(self):
        """
        Display the solved board or error message if no solution exists
        """
        
        if self.solve(0, 0):
            print("\nSolution exists")
            for i in range(self.M):
                for j in range(self.N):
                    print(self.board[i][j], end=' ')
                print()
        else:
            print("Solution does not exist")


if __name__ == '__main__':
    try:
        print("Dimensions must be such that there are even number of cells")
        M, N = list(map(int, input('Enter the board dimensions : ').split())) 
        if M*N % 2 != 0:
            raise Exception
        
        print('Enter -1 if any number of characters allowed')
        top = list(map(int, input('Enter "+" for columns : ').split()))
        if len(top) != N:
            raise Exception
        
        bottom = list(map(int, input('Enter "-" for columns : ').split()))
        if len(bottom) != N:
            raise Exception

        left = list(map(int, input('Enter "+" for rows : ').split()))
        if len(left) != M:
            raise Exception

        right = list(map(int, input('Enter "-" for rows : ').split()))
        if len(right) != M:
            raise Exception

        slots = [list(map(str, input('Input slots corresponding to row ' + str(i+1) + ": ").split())) for i in range(M)]
        if len(slots) != M:
            raise Exception
        for i in range(len(slots)):
            if len(slots[i]) != N:
                raise Exception
    except Exception:
        print("\nUsing default values since wrong input was given\n")
        (M, N) = (5, 6)
        top = [3, -1, -1, 2, 1, -1]
        bottom = [2, -1, -1, 2, -1, 3]
        left = [2, 3, -1, -1, -1]
        right = [-1, -1, -1, 1, -1]
        slots = [
            ['L', 'R', 'L', 'R', 'T', 'T'],
            ['L', 'R', 'L', 'R', 'B', 'B'],
            ['T', 'T', 'T', 'T', 'L', 'R'],
            ['B', 'B', 'B', 'B', 'T', 'T'],
            ['L', 'R', 'L', 'R', 'B', 'B']
        ]
        
    print("Top : ", top)
    print("Bottom : ", bottom)
    print("Left : ", left)
    print("Right : ", right)
    print("Slots given : ", slots)
    ob = MagnetPuzzle((M, N), top, bottom, left, right, slots)
    ob.display()

File: cpp/backtracking/Magnet_Puzzle.cpp
//Magnet Puzzle
#include <iostream>
using namespace std;
 
// M x N matrix
#define M 5
#define N 6
 
// A utility function to print solution
void printSolution(char board[M][N])
{
    for (int i = 0; i < M; i++)
    {
        for (int j = 0; j < N; j++)
            cout << board[i][j] << " ";
        cout << endl;
    }
}
 
// A utility function to count number of characters ch in current column j
int countInColumns(char board[N][N], char ch, int j)
{
    int count = 0;
    for (int i = 0; i < M; i++)
        if (board[i][j] == ch)
            count++;
 
    return count;
}
 
// A utility function to count number of characters ch in current row i
int countInRow(char board[N][N], char ch, int i)
{
    int count = 0;
    for (int j = 0; j < N; j++)
        if (board[i][j] == ch)
            count++;
 
    return count;
}
 
// Function to check if it safe to put 'ch' at board[row][col]
bool isSafe(char board[N][N], int row, int col, char ch, int top[],
             int left[], int bottom[], int right[])
{
    // check for adjacent cells
    if ((row - 1 >= 0 && board[row - 1][col] == ch) ||
        (col + 1 < N && board[row][col + 1] == ch) ||
        (row + 1 < M && board[row + 1][col] == ch) ||
        (col - 1 >= 0 && board[row][col - 1] == ch))
    return false;
 
    // count ch in current row
    int rowCount = countInRow(board, ch, row);
 
    // count ch in current column
    int colCount = countInColumns(board, ch, col);
 
    // if given character is '+', check top[] & left[]
    if (ch == '+')
    {
        // check top
        if (top[col] != -1 && colCount >= top[col])
            return false;
 
        // check left
        if (left[row] != -1 && rowCount >= left[row])
            return false;
    }
 
    // if given character is '-', check bottom[] & right[]
    if (ch == '-')
    {
        // check bottom
        if (bottom[col] != -1 && colCount >= bottom[col])
            return false;
 
        // check left
        if (right[row] != -1 && rowCount >= right[row])
            return false;
    }
 
    return true;
}
 
// Function to validate Configuration of output board
bool validateConfiguration(char board[N][N], int top[], int left[],
                        int bottom[], int right[])
{
    // check top
    for (int i = 0; i < N; i++)
        if (top[i] != -1 &&
            countInColumns(board, '+', i) != top[i])
        return false;
 
    // check left
    for (int j = 0; j < M; j++)
        if (left[j] != -1 &&
            countInRow(board, '+', j) != left[j])
        return false;
 
    // check bottom
    for (int i = 0; i < N; i++)
        if (bottom[i] != -1 &&
            countInColumns(board, '-', i) != bottom[i])
        return false;
 
    // check right
    for (int j = 0; j < M; j++)
        if (right[j] != -1 &&
            countInRow(board, '-', j) != right[j])
        return false;
 
    return true;
}
 
// Main function to solve the Bipolar Magnets puzzle
bool solveMagnetPuzzle(char board[N][N], int row, int col, int top[],
                int left[], int bottom[], int right[], char str[M][N])
{
    // if we have reached last cell
    if (row >= M - 1 && col >= N - 1)
    {
        if (validateConfiguration(board, top, left, bottom, right))
            return true;
 
        return false;
    }
 
    // if last column of current row is already processed,
    // go to next row, first column
    if (col >= N)
    {
        col = 0;
        row = row + 1;
    }
 
    // if current cell contains 'R' or 'B' (end of horizontal
    // or vertical slot) recur for next cell
    if (str[row][col] == 'R' || str[row][col] == 'B')
    {
        if (solveMagnetPuzzle(board, row, col + 1, top,
                                left, bottom, right, str))
            return true;
    }
 
    // if horizontal slot contains L and R
    if (str[row][col] == 'L' && str[row][col + 1] == 'R')
    {
        // put ('+', '-') pair and recur
        if (isSafe(board, row, col, '+', top, left, bottom, right) &&
            isSafe(board, row, col + 1, '-', top, left, bottom, right))
        {
            board[row][col] = '+';
            board[row][col + 1] = '-';
 
            if (solveMagnetPuzzle(board, row, col + 2,
                            top, left, bottom, right, str))
                return true;
 
            // if it doesn't lead to a solution, backtrack
            board[row][col] = 'X';
            board[row][col + 1] = 'X';
        }
 
        // put ('-', '+') pair and recur
        if (isSafe(board, row, col, '-', top, left, bottom, right) &&
            isSafe(board, row, col + 1, '+', top, left, bottom, right))
        {
            board[row][col] = '-';
            board[row][col + 1] = '+';
 
            if (solveMagnetPuzzle(board, row, col + 2,
                            top, left, bottom, right, str))
                return true;
 
            // if it doesn't lead to a solution, backtrack
            board[row][col] = 'X';
            board[row][col + 1] = 'X';
        }
    }
 
    // if vertical slot contains T and B
    if (str[row][col] == 'T' && str[row + 1][col] == 'B')
    {
        // put ('+', '-') pair and recur
        if (isSafe(board, row, col, '+', top, left, bottom, right) &&
            isSafe(board, row + 1, col, '-', top, left, bottom, right))
        {
            board[row][col] = '+';
            board[row + 1][col] = '-';
 
            if (solveMagnetPuzzle(board, row, col + 1,
                            top, left, bottom, right, str))
                return true;
 
            // if it doesn't lead to a solution, backtrack
            board[row][col] = 'X';
            board[row + 1][col] = 'X';
        }
 
        // put ('-', '+') pair and recur
        if (isSafe(board, row, col, '-', top, left, bottom, right) &&
            isSafe(board, row + 1, col, '+', top, left, bottom, right))
        {
            board[row][col] = '-';
            board[row + 1][col] = '+';
 
            if (solveMagnetPuzzle(board, row, col + 1,
                            top, left, bottom, right, str))
                return true;
 
            // if it doesn't lead to a solution, backtrack
            board[row][col] = 'X';
            board[row + 1][col] = 'X';
        }
    }
 
    // ignore current cell and recur
    if (solveMagnetPuzzle(board, row, col + 1,
                        top, left, bottom, right, str))
        return true;
 
    // if no solution is possible, return false
    return false;
}
 
void solveMagnetPuzzle(int top[], int left[], int bottom[],
                        int right[], char str[M][N])
{
    // to store result
    char board[M][N];
 
    // initialize all cells by 'X'
    for (int i = 0; i < M; i++)
        for (int j = 0; j < N; j++)
            board[i][j] = 'X';
 
    // start from (0, 0) cell
    if (!solveMagnetPuzzle(board, 0, 0, top, left, bottom, right, str))
    {
        cout << "Solution does not exist";
        return;
    }
 
    // print result if given configuration is solvable
    printSolution(board);
}
 
int main()
{
    // indicates the count of + or - along the top (+), bottom (-),
    // left (+) and right (-) edges respectively.
    // Value of -1 indicates any number of + or - signs
    int top[N] = { 1, -1, -1, 2, 1, -1 };
    int bottom[N] = { 2, -1, -1, 2, -1, 3 };
    int left[M] = { 2, 3, -1, -1, -1 };
    int right[M] = { -1, -1, -1, 1, -1 };
 
    // rules matrix
    char str[M][N] =
    {
        { 'L', 'R', 'L', 'R', 'T', 'T' },
        { 'L', 'R', 'L', 'R', 'B', 'B' },
        { 'T', 'T', 'T', 'T', 'L', 'R' },
        { 'B', 'B', 'B', 'B', 'T', 'T' },
        { 'L', 'R', 'L', 'R', 'B', 'B' }
    };
 
    solveMagnetPuzzle(top, left, bottom, right, str);
 
    return 0;
}
